pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -966,6 +977,45 @@ func (ctrl *PersistentVolumeController) assignDefaultStorageClass(ctx context.Co
 	return true, nil
 }
 
+// assignDefaultVolumeAttributesClass updates the claim volume attributes class if there is any, the claim is updated to the API server.
+// Ignores claims that already have a volume attributes class.
+// TODO: if resync is ever changed to a larger period, we might need to change how we set the default volume attributes class on existing claims
+func (ctrl *PersistentVolumeController) assignDefaultVolumeAttributesClass(ctx context.Context, claim *v1.PersistentVolumeClaim) (bool, error) {
+	logger := klog.FromContext(ctx)
+
+	if claim.Spec.VolumeAttributesClassName != nil {
+		return false, nil
+	}
+
+	className := storagehelpers.GetPersistentVolumeClaimClass(claim)
+	if className == "" {
+		return false, nil
+	}
+	sc, err := ctrl.classLister.Get(className)
+	if err != nil {
+		return false, err
+	}
+
+	vac, err := util.GetDefaultVolumeAttributesClass(ctrl.vacLister, sc.Provisioner)
+	if err != nil {
+		return false, err
+	} else if vac == nil {
+		logger.V(4).Info("Can not assign volume attributes class to PersistentVolumeClaim: default volume attributes class not found", "driverName", sc.Provisioner, "PVC", klog.KObj(claim))
+		return false, nil
+	}
+
+	logger.V(4).Info("Assigning VolumeAttributesClass to PersistentVolumeClaim", "PVC", klog.KObj(claim), "volumeAttributesClassName", vac.Name)
+	claim.Spec.VolumeAttributesClassName = &vac.Name
+
+	_, err = ctrl.kubeClient.CoreV1().PersistentVolumeClaims(claim.GetNamespace()).Update(ctx, claim, metav1.UpdateOptions{})
```

Comment:
What if we have already started to provision the volume. Do we want to change it in the middle of the operation? @sunnylovestiramisu

Reply:
I am thinking we should still allow it, but depending on the provider, some of them have restriction of how often can you make a update volume call under the hood, and the request will get rejected if it is called twice within the limited time.

Reply:
Discussed today in csi standup.

The defaulting of VAC should only happen if the PVC.vacName is nil. Once it is set, PV controller should not change the VAC again.

There is another question on what happens if the user changes the PVC.vacName in the middle of provisioning. We can discuss more on external-provisioner PR how we should handle that.